src/plugins/intel_cpu/src/nodes/mkldnn_roi_align_node.cpp

```diff
@@ -15,14 +15,631 @@
+    Xmm xmm_dst_tail = Xmm(vmm_dst_tail.getIdx());
+
+    Vmm vmm_temp1 = vmm_weights1;
+    Xmm xmm_temp1 = Xmm(vmm_temp1.getIdx());
+    Vmm vmm_temp2 = vmm_weights2;
+    Xmm xmm_temp2 = Xmm(vmm_temp2.getIdx());
+
+    Vmm vmm_mask = vmm_weights3;
+
+    Opmask k_mask = Opmask(7);
+
+    reg64_t reg_params = abi_param1;
+
+    void roi_align_cgather() {
+        mov(reg_src_address, ptr[reg_params + GET_OFF(src)]);
+        mov(reg_weights, ptr[reg_params + GET_OFF(weights)]);
+
+        mov(reg_num_samples, ptr[reg_params + GET_OFF(num_samples)]);
+        mov(reg_work_amount, ptr[reg_params + GET_OFF(work_amount)]);
+        if (jcp_.alg == Algorithm::ROIAlignAvg) {
+            mov(reg_tmp_64, ptr[reg_params + GET_OFF(scale)]);
+            uni_vbroadcastss(vmm_scale, ptr[reg_tmp_64]);
+        }
+        mov(reg_tmp_64, ptr[reg_params + GET_OFF(buffer)]);
+        uni_vpinsrq(xmm_args_pool, xmm_args_pool, reg_tmp_64, 0);
+        mov(reg_tmp_64, ptr[reg_params + GET_OFF(dst)]);
+        uni_vpinsrq(xmm_args_pool, xmm_args_pool, reg_tmp_64, 1);
+
+        if (jcp_.layout == ROIAlignLayoutType::nspc) {
+            int src_stride = v_step * jcp_.data_size;
+            mov(reg_src_stride, src_stride);
+        } else if (jcp_.layout == ROIAlignLayoutType::blk) {
+            mov(reg_src_stride, ptr[reg_params + GET_OFF(src_stride)]);
+            imul(reg_src_stride, reg_src_stride, jcp_.data_size);
+        }
+
+        auto store = [&](Vmm vmm_dst, Xbyak::Reg64 reg_dst, int elt_num) {
+                    store_emitter->emit_code({static_cast<size_t>(vmm_dst.getIdx())}, {static_cast<size_t>(reg_dst.getIdx())},
+                    std::make_shared<store_emitter_context>(Precision::FP32, jcp_.data_prc, elt_num),
+                    {store_pool_vec_idxs}, {store_pool_gpr_idxs});
+        };
+
+        auto load_buf = [&](Xbyak::Reg64 reg_src, Vmm vmm_src, int elt_num) {
+                    load_emitter->emit_code({static_cast<size_t>(reg_src.getIdx())}, {static_cast<size_t>(vmm_src.getIdx())},
+                    std::make_shared<load_emitter_context>(Precision::FP32, Precision::FP32, elt_num),
+                    {}, {load_pool_gpr_idxs});
+        };
+        auto store_buf = [&](Vmm vmm_dst, Xbyak::Reg64 reg_dst, int elt_num) {
+                    store_emitter->emit_code({static_cast<size_t>(vmm_dst.getIdx())}, {static_cast<size_t>(reg_dst.getIdx())},
+                    std::make_shared<store_emitter_context>(Precision::FP32, Precision::FP32, elt_num),
+                    {store_pool_vec_idxs}, {store_pool_gpr_idxs});
+        };
+
+        // out loop for samples in bin
+        Xbyak::Label out_loop_label;
+        Xbyak::Label out_loop_end_label;
+
+        mov(reg_src_oc, 0);
+        mov(reg_weights_oc, 0);
+
+        L(out_loop_label);
+        {
+            cmp(reg_num_samples, 1);
+            jl(out_loop_end_label, T_NEAR);
+
+            // get 4 src address and 4 vmm_weights
+            get_src(reg_src0);
```

Comment:
BTW, why do you need 4 separate pointers given they are always placed sequentially in the memory. In other words you can have one pointer and const offset for the others,

Reply:
The address of src address is sequentially, but the src address itself is not sequentially. We need 4 src to perf bilinear interpolate and need advance the 4 src address independently to C.